tests/fake/test_fake_publish.py

```diff
@@ -63,3 +73,52 @@ def test_publish_absent_raises():
     exception = repo_copy2.publish().exception()
     assert isinstance(exception, PulpException)
     assert "repo1 not found" in str(exception)
+
+
+def test_can_upload(tmpdir):
+    """repo.upload_file() succeeds with fake client and populates upload_history."""
+    controller = FakeController()
+
+    controller.insert_repository(FileRepository(id="repo1"))
+
+    client = controller.client
+    repo1 = client.get_repository("repo1").result()
+
+    somefile = tmpdir.join("some-file.txt")
+    somefile.write(b"there is some binary data:\x00\x01\x02")
+
+    upload_f = repo1.upload_file(str(somefile))
+
+    # The future should resolve successfully
+    tasks = upload_f.result()
+
+    # It should have returned at least one successful task.
+    assert tasks
+    for task in tasks:
+        assert task.succeeded
+
+    # The change should be reflected in the controller's publish history
+    history = controller.upload_history
+
+    assert len(history) == 1
+    assert history[0].repo_id == "repo1"
+    assert history[0].tasks == tasks
+
+
+def test_upload_nonexistent_file_raises():
+    """repo.upload_file() with nonexistent file fails with fake client"""
+    controller = FakeController()
+
+    controller.insert_repository(FileRepository(id="repo1"))
+
+    client = controller.client
+    repo1 = client.get_repository("repo1").result()
+
+    # If file's not found, Python 2 raises IOError and Python 3 raises
+    # FileNotFoundError. The latter one is not defined in Python 2.
+    if sys.version_info < (3,):
+        exception = IOError
+    else:
+        exception = FileNotFoundError
+    with pytest.raises(exception):
+        upload_f = repo1.upload_file("nonexistent_file")
```

Comment:
I think this is actually pointing to a difference between the real and fake clients.

Question: if you use the real client and you attempt to upload a file which doesn't exist, does it:

* (A) - raise an exception, or
* (B) - return a future which raises an exception

I think it might be (B), but you've made the fake client do (A).

Can you please investigate and ensure they behave the same?

Reply:
make sense.

Reply:
Hm, actually, it just raises an exception directly, because open the file isn't wrapped in the a future, see code:
In file.py, upload_file calls _do_upload_file as:
`upload_complete_f = self._client._do_upload_file(upload_id, file_obj, name)`
in client.py, it opens the file first, then wrap the upload in future and return:
`is_file_object = "close" in dir(file_obj)
        if not is_file_object:
            file_obj = open(file_obj, "rb")

        LOG.info("Uploading %s to Pulp", name)
        upload_f = f_flat_map(f_return(), lambda _: do_next_upload(hashlib.sha256(), 0))`

So, this is not completely non-blocking, try to upload the file not existed will fail the whole process, is this OK?

Reply:
I am fine with either way as long as both the clients are consistent.